controllers/ingress_controller.go

```diff
@@ -80,102 +116,142 @@ func (r *IngressReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 
 	// /FINALIZER LOGIC ================================================================================================
 
-	if desired.Spec.IngressClassName == nil {
-		log.Info("IngressClassName should not be nil")
-		return ctrl.Result{}, errors.New("IngressClassName should not be nil")
+	// ensure there is a template api definition
+	templateAnnotationValue, ok := desired.Annotations[ingressTemplateAnnotationKey]
+	if !ok {
+		return ctrl.Result{}, fmt.Errorf("expecting template annotation %s", ingressTemplateAnnotationKey)
 	}
 
-	ingressClassList := &v1beta1.IngressClassList{}
-	opts := []client.ListOption{
-		client.InNamespace(req.Namespace),
-	}
-	err := r.List(ctx, ingressClassList, opts...)
+	// we have parameters - as such, we should ensure that there is an api definition resource
+	template := &v1alpha1.ApiDefinition{}
+	err := r.Get(ctx, types.NamespacedName{Name: templateAnnotationValue, Namespace: req.Namespace}, template)
 	if err != nil {
-		println(err.Error())
+		log.Error(err, "error getting api definition to use as a template", "name", templateAnnotationValue)
 		return ctrl.Result{}, err
 	}
 
-	println(ingressClassList.Kind)
+	apisToCreateOrUpdate := v1alpha1.ApiDefinitionList{}
+	apisToUpdate := v1alpha1.ApiDefinitionList{}
+	apisToCreate := v1alpha1.ApiDefinitionList{}
+	apisToDelete := v1alpha1.ApiDefinitionList{}
 
-	ingressClass := &v1beta1.IngressClass{}
-	if err := r.Get(ctx, types.NamespacedName{Name: *desired.Spec.IngressClassName, Namespace: req.Namespace}, ingressClass); err != nil {
-		println(err.Error())
-		return ctrl.Result{}, err
-	}
+	for _, rule := range desired.Spec.Rules {
+		hostName := rule.Host
+		for _, p := range rule.HTTP.Paths {
+			apiName := strings.TrimRight(strings.ReplaceAll(fmt.Sprintf("%s-%s-%s-%s", req.Name, req.Namespace, hostName, p.Path), "/", ""), "-")
```

Comment:
Anybody got any ideas for a decent API Naming convention?

Reply:
can we move this to a separate function first? If I understand correctly we will be deriving the names here  for both creating and updates so this function should be able to generate unique names for each Ingress rule.